plat/renesas/rcar/drivers/board/board.c

```diff
@@ -27,83 +27,104 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
-
 #include <stdint.h>
 #include <iic_dvfs.h>
 #include "board.h"
 
+static uint8_t board_id = BOARD_ID_UNKNOWN;
 
-/************************************************************************
- * Defines
- ************************************************************************/
-#ifndef BOARD_DEFAULT
-#define BOARD_DEFAULT		(BOARD_SALVATOR_X << BOARD_CODE_SHIFT)
+void board_id_init()
+{
+	uint8_t id = BOARD_ID_UNKNOWN;
+#if PMIC_ON_BOARD && !defined(BOARD_ID)
+	int ret;
 #endif
 
-#define SLAVE_ADDR_EEPROM	(0x50U)
-#define	REG_ADDR_BOARD_ID	(0x70U)
+	/* board detection and validation had been done already */
+	if (board_id != BOARD_ID_UNKNOWN)
+		return;
 
-#define BOARD_CODE_MASK		(0xF8U)
-#define BOARD_REV_MASK		(0x07U)
-#define BOARD_CODE_SHIFT	(3U)
+#if PMIC_ON_BOARD && !defined(BOARD_ID)
+	/* Board ID detection from EEPROM */
+	ret = rcar_iic_dvfs_recieve(PMIC_EEPROM_SLAVE_ADDR,
+				PMIC_EEPROM_BOARD_ID_REG_ADDR, &id);
+	if (ret != 0)
+		id = BOARD_ID_UNKNOWN;
+#endif
 
-#define BOARD_ID_UNKNOWN	(0xFFU)
+#ifdef BOARD_ID
+	id = BOARD_ID;
+#endif
+	if (id == BOARD_ID_UNKNOWN)
+		id = BOARD_ID_DEFAULT;
 
+	switch (id >> BOARD_ID_NAME_SHIFT) {
+	case BOARD_SALVATOR_X:
+	case BOARD_KRIEK:
+	case BOARD_STARTER_KIT:
+	case BOARD_SALVATOR_XS:
+		break;
+	default:
+		id = BOARD_ID_DEFAULT;
+		break;
+	}
+
+	board_id = id;
+}
 
-/************************************************************************
- * Global variables
- ************************************************************************/
-const char *g_board_tbl[] = {
+static const char *board_name[] = {
 	[BOARD_SALVATOR_X]	= "Salvator-X",
-	[BOARD_SALVATOR_XS]	= "Salvator-XS",
 	[BOARD_KRIEK]		= "Kriek",
-	[BOARD_STARTER_KIT]	= "Starter Kit"
+	[BOARD_STARTER_KIT]	= "Starter Kit",
+	[BOARD_SALVATOR_XS]	= "Salvator-XS"
 };
-const char *g_board_unknown	= "unknown";
 
+int32_t get_board_id()
+{
+	return board_id;
+}
 
-int32_t get_board_type(uint32_t *type, uint32_t *rev)
+const char *get_board_name()
 {
-	int32_t ret = 0;
-	uint8_t read_rev;
-	static uint8_t g_board_id = BOARD_ID_UNKNOWN;
-	const uint8_t board_tbl[][8U] = {
-		[BOARD_SALVATOR_X]	= {0x10U, 0x11U, 0xFFU, 0xFFU,
-					   0xFFU, 0xFFU, 0xFFU, 0xFFU},
-		[BOARD_KRIEK]		= {0x10U, 0xFFU, 0xFFU, 0xFFU,
-					   0xFFU, 0xFFU, 0xFFU, 0xFFU},
-		[BOARD_STARTER_KIT]	= {0x10U, 0xFFU, 0xFFU, 0xFFU,
-					   0xFFU, 0xFFU, 0xFFU, 0xFFU},
-		[BOARD_SALVATOR_XS]	= {0x10U, 0xFFU, 0xFFU, 0xFFU,
-					   0xFFU, 0xFFU, 0xFFU, 0xFFU},
-	};
-
-	if (BOARD_ID_UNKNOWN == g_board_id) {
-#if PMIC_ON_BOARD
-		/* Board ID detection from EEPROM */
-		ret = rcar_iic_dvfs_recieve(SLAVE_ADDR_EEPROM,
-			REG_ADDR_BOARD_ID, &g_board_id);
-		if (0 != ret) {
-			g_board_id = BOARD_ID_UNKNOWN;
-		} else if (BOARD_ID_UNKNOWN == g_board_id) {
-			/* Can't recognize the board */
-			g_board_id = BOARD_DEFAULT;
-		} else {
-			/* none */
-		}
-#else
-		g_board_id = BOARD_DEFAULT;
-#endif
-	}
+	return board_name[board_id];
```

Comment:
please note the bug in get_board_name. the board name id is (board_id >> BOARD_NAME_ID_SHIFT).